Fixed merging of host's and dns' search lines

vefimova opened the pull request:
Fixed forming of pod's Search line in resolv.conf:
 - exclude duplicates while merging of host's and dns' search lines to form pod's one
 - truncate pod's search line if it exceeds resolver limits: is > 255 chars and containes > 6 searches
 - monitoring the resolv.conf file which is used by kubelet (set thru --resolv-conf="") and logging and eventing if search line in it consists of more than 3 entries (or 6 if Cluster Domain is set) or its lenght is > 255 chars
 - logging and eventing when a pod's search line is > 255 chars or containes > 6 searches during forming

Fixes #29270

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Fixed forming resolver search line for pods: exclude duplicates, obey libc limitations, logging and eventing appropriately.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29666)

<!-- Reviewable:end -->

vefimova committed:
Fixed forming of pod's Search line in resolv.conf:
 - exclude duplicates while merging of host's and dns' search lines to form pod's one
 - truncate pod's search line if it exceeds resolver limits: is > 255 chars and containes > 6 searches
 - monitoring the resolv.conf file which is used by kubelet (set thru --resolv-conf="") and logging and eventing if search line in it consists of more than 3 entries
   (or 6 if Cluster Domain is set) or its lenght is > 255 chars
 - logging and eventing when a pod's search line is > 255 chars or containes > 6 searches during forming
Fixes #29270

dims commented:
@vefimova Can we please log at the debug level when we find duplicates and/or drop things on the floor? Apologize if we do this already somewhere

vishh commented on pkg/kubelet/types/constants.go:
```diff
@@ -19,4 +19,7 @@ package types
 	ResolvConfDefault = "/etc/resolv.conf"
+	// resolver file Search line current limitations
+	MaxDomains = 6
+	MaxLen     = 255
```
Max len of what?

vefimova replied:
fixed

vishh commented on pkg/kubelet/types/constants.go:
```diff
@@ -19,4 +19,7 @@ package types
 	// system default DNS resolver configuration
 	ResolvConfDefault = "/etc/resolv.conf"
+	// resolver file Search line current limitations
+	MaxDomains = 6
```
nit: Qualify the constant name. `MaxDNSDomains`

vefimova replied:
fixed

vishh commented:
This PR seems to be technically correct, but I wonder what the consequences will be. Silently dropping search paths might result in subtle issues in production. Does it make more sense to fail pods rather than starting them with potentially unacceptable dns config?

vefimova commented:
@vishh 
AFAIS, if to return error in such case then the pod itself will be reported as created, but the container won't be started due to error we return, let's say: "dns search string exceeds limits". 
But in such case, what should the end user do? Dns search formed by kubelet always consists of 3 domains. Next, + host dns domains. So what is the option in case of error? Don't use cluster DNS? Probably, user isin't supposed to have access to the node to change node's resolv.conf and shouldn't do it.

Also I've check glibc code parsing search line, it just silently cut off string to 256 chars and then parse until 6 domains will be found:
https://sourceware.org/git/?p=glibc.git;a=blob;f=resolv/res_init.c;h=cea4c8a3cc5ad0181ee706daedb940083d08ba76;hb=b898b64be7957cf2a06a165e6ba57fa067fc5685 (see L225 and L236)

So I suppose there should be no harm. I propose just to add warning logs as @dims suggested in case we perform some editing of search line. WDYT?

vishh commented:
Logs are not easily accessible to users. Subtle failures are much harder to
track. Given that the search paths are pre-computed for pods, why can't the
kubelet fail a pod without even attempting to start containers?

On Fri, Aug 5, 2016 at 8:40 AM, Victoria Efimova <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh>
> AFAIS, if to return error in such case then the pod itself will be
> reported as created, but the container won't be started due to error we
> return, let's say: "dns search string exceeds limits".
> But in such case, what should the end user do? Dns search formed by
> kubelet always consists of 3 domains. Next, + host dns domains. So what is
> the option in case of error? Don't use cluster DNS? Probably, user isin't
> supposed to have access to the node to change node's resolv.conf and
> shouldn't do it.
>
> Also I've check glibc code parsing search line, it just silently cut off
> string to 256 chars and then parse until 6 domains will be found:
> https://sourceware.org/git/?p=glibc.git;a=blob;f=resolv/res_init.c;h=
> cea4c8a3cc5ad0181ee706daedb940083d08ba76;hb=b898b64be7957cf2a06a165e6ba57f
> a067fc5685 (see L225 and L236)
>
> So I suppose there should be no harm. I propose just to add warning logs
> as @dims <https://github.com/dims> suggested in case we perform some
> editing of search line. WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29666#issuecomment-237884683>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKDezN5C2kqdBvfpz9RjmobQbC1kHks5qc1lygaJpZM4JV-Xz>
> .
>

vefimova commented:
But still what do you expect from user in such case? How he is supposed to fix pod definition?

Also that patch doesn't change existent behaviour, i.e. if there is no conflicts, it will just provide valid docker content to work with. As resolver has this limits hardcoded, but also doesn't notify about invalid resolv.conf content.

Another case with duplicates, it's just a bug, and should be fixed to thow away them from search line.

vefimova commented:
@vishh @thockin @girishkalele 

So guys, do you still have concerns?
Actually, we have the same approach already in code for handling the len of hostname: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1180-L1183

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -121,6 +122,64 @@ func (kl *Kubelet) parseResolvConf(reader io.Reader) (nameservers []string, sear
 }
 
+func appendAndContinue(composedSearch []string, domain string) bool {
+	searchLen := len(strings.Join(composedSearch, " "))
```
len(composedSearch)_+ 1

thockin replied:
nevermind, I am dumb

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -121,6 +122,64 @@ func (kl *Kubelet) parseResolvConf(reader io.Reader) (nameservers []string, sear
+func appendAndContinue(composedSearch []string, domain string) bool {
+	searchLen := len(strings.Join(composedSearch, " "))
+
+	if searchLen == 0 || (searchLen+len(domain)+1 <= types.MaxSearchLineLen && len(composedSearch) < types.MaxSearchLineDNSDomains) {
```
==0 is impossible

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -121,6 +122,64 @@ func (kl *Kubelet) parseResolvConf(reader io.Reader) (nameservers []string, sear
 	return parseResolvConf(reader, scrubber)
 }
 
+func appendAndContinue(composedSearch []string, domain string) bool {
```
can you name this in a way that doesn't suggest it is doing the append?  `wouldFit`  or something

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -121,6 +122,64 @@ func (kl *Kubelet) parseResolvConf(reader io.Reader) (nameservers []string, sear
+			glog.V(8).Infof("Found and omitted duplicated dns domain in host search line: '%s' during merging with cluster dns domains", hostDomain)
+			continue
+		}
+		if appendAndContinue(composedSearch, hostDomain) {
```
capture L160-L165 as a function, since it is duplicated above.

```
composedSearch, err := appendOrError(composedSearch, hostDomain)
if err != nil {
  glog...
  return
```

thockin commented on pkg/kubelet/kubelet_network_test.go:
```diff
@@ -136,6 +136,44 @@ func TestParseResolvConf(t *testing.T) {
+		hostNames    []string
+		resultSearch []string
+	}{
+		{[]string{"default.svc.cluster.local", "svc.cluster.local", "cluster.local"},
```
the linewrapping makes these hard to visually compare.  Please put wrap at the open/close curlies for the struct

thockin replied:
and for the individual struct members

thockin replied:
All of these would be easier to read it you just used AAA, BBB, CCC for names

thockin commented on pkg/kubelet/kubelet_network_test.go:
```diff
@@ -136,6 +136,44 @@ func TestParseResolvConf(t *testing.T) {
+		{[]string{"default.svc.cluster.local", "svc.cluster.local", "cluster.local"},
+			[]string{}, []string{"default.svc.cluster.local", "svc.cluster.local", "cluster.local"}},
+
+		{[]string{"default.svc.cluster.local", "svc.cluster.local", "cluster.local"},
```
add a simple case with A, B + C, D = ABCD

thockin commented on pkg/kubelet/kubelet_network_test.go:
```diff
@@ -136,6 +136,44 @@ func TestParseResolvConf(t *testing.T) {
+
+		{[]string{"cluster.local"},
+			[]string{"cluster.local",
+				"3.141592653589793238462643383279502884197169399375105820974944592.com",
```
strings.Repeat("a", 100) makes this WAY easier to read

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -121,6 +122,64 @@ func (kl *Kubelet) parseResolvConf(reader io.Reader) (nameservers []string, sear
+			}
+		}
+		if !unique {
+			glog.V(8).Infof("Found and omitted duplicated dns domain in host search line: '%s' during merging with cluster dns domains", hostDomain)
```
I would legit log these.  glog.Errorf()

thockin commented:
Whether we log and truncate or actually let the bad data pass through is debatable, but this is safer.  You could argue for a node-related or pod-related event here.  What do you think Vish? 

Thanks for the PR

vishh commented:
There are two scenarios to consider:

1. The number of search paths on the node is > 3, which will result in kubelet silently dropping some of them inside of the container (thanks @thockin for identifying this scenario)
1. The length of the namespace name of a pod which could lead to dropping of one or more search paths inside the container

I prefer failing explicitly over logging and moving on. 
Specifically, the node should constantly monitor the number of search paths on the host and if it exceeds `3`, it should mark itself as unhealthy `NotReady`. Pods will automatically be re-scheduled in this case.
If a user updates the search path on all the nodes, then they would have bricked their cluster. But should they be rolling out all at once?
If the namespace name is too long and it results in dropping of search paths, the pod should ideally be failed, an event generated for the pod indicating search paths are the reason for the failure. What would the user do in this case? They either reduce the length of their namespace name(s) and/or  cluster domain, and/or host search paths.

In theory, this PR is only improving today's situation where we blindly truncate search paths. But given that we are attempting to fix it, can we also include the changes mentioned above?

thockin commented:
While I agree that "fail obviously" is powerful, the failure mode of too
many search path entries is almost certainly not catastrophic enough to
warrant going full meltdown.

I also understand the idea that a too-long namespace might cause pods in
that namespace to fail to start.  But again, this can change underneath the
user.  An admin (or a troublemaker) could change the search path to be
close to the limits, such that running pods are fine, but when they crash
they will fail to restart.

So I think we have to consider the severity of the failure mode, and I am
not sure I buy that this failure warrants a cluster bomb (hah!).

A goroutine that monitors the flag-configured resolv.conf for > 3 search
paths and logs an event against the node is uncontroversial.

Logging and eventing when a pod  is > 255 or > 6 searches is
uncontroversial.

Failing is controversial.  Can we do the non-scary parts first?



On Tue, Aug 16, 2016 at 5:27 PM, Vish Kannan <notifications@github.com>
wrote:

> There are two scenarios to consider:
>
>    1. The number of search paths on the node is > 3, which will result in
>    kubelet silently dropping some of them inside of the container (thanks
>    @thockin <https://github.com/thockin> for identifying this scenario)
>    2. The length of the namespace name of a pod which could lead to
>    dropping of one or more search paths inside the container
>
> I prefer failing explicitly over logging and moving on.
> Specifically, the node should constantly monitor the number of search
> paths on the host and if it exceeds 3, it should mark itself as unhealthy
> NotReady. Pods will automatically be re-scheduled in this case.
> If a user updates the search path on all the nodes, then they would have
> bricked their cluster. But should they be rolling out all at once?
> If the namespace name is too long and it results in dropping of search
> paths, the pod should ideally be failed, an event generated for the pod
> indicating search paths are the reason for the failure. What would the user
> do in this case? They either reduce the length of their namespace name(s)
> and/or cluster domain, and/or host search paths.
>
> In theory, this PR is only improving today's situation where we blindly
> truncate search paths. But given that we are attempting to fix it, can we
> also include the changes mentioned above?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29666#issuecomment-240280041>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVMNF0VMJ6BFqkRyY0r2NHlXeoBo2ks5qglV4gaJpZM4JV-Xz>
> .
>

vishh commented:
@thockin

Agreed. Perhaps, we can provide an opt-out of the failing pods behavior for
unreasonably long namespace name to ease with the transition? Again, we are
considering the remote corner case hopefully.

On Mon, Aug 29, 2016 at 5:37 PM, Kubernetes Bot <notifications@github.com>
wrote:

> Can one of the admins verify that this patch is reasonable to test? If so,
> please reply "ok to test".
> (Note: "add to whitelist" is no longer supported. Please update
> configurations in kubernetes/test-infra/jenkins/job-configs/kubernetes-
> jenkins-pull
> <https://github.com/kubernetes/test-infra/tree/master/jenkins/job-configs/kubernetes-jenkins-pull>
> instead.)
>
> This message will repeat several times in short succession due to
> jenkinsci/ghprb-plugin#292
> <https://github.com/jenkinsci/ghprb-plugin/issues/292>. Sorry.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29666#issuecomment-243299696>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKDNEYxARj1JHP7cSHSSTx_tNjME5ks5qk3thgaJpZM4JV-Xz>
> .
>

vefimova commented:
@thockin @vishh Hi guys, sorry for delay, wiil provide patch with proposed approach soon. 
To sum up what need to be done:
- monitoring the resolv.conf file which is used during pod creating (set thru --resolv-conf="") and logging if search line in it consists of more than 3 entries.
- Logging and eventing when a pod  is > 255 or > 6 searches

thockin commented:
Ping @vefimova are you still on this?  It's a good PR...

vefimova commented:
@thockin yes, will update it soon.

vefimova commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/kubelet/kubelet_network.go, line 125 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7d-r2-125:-KZ2LN3hDz01AU0CQ7Rj:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network.go#L125)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

can you name this in a way that doesn't suggest it is doing the append?  `wouldFit`  or something
</blockquote></details>

Done.

---

*[pkg/kubelet/kubelet_network.go, line 157 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7d-r2-157:-KZ2LO23cTJSxUemW1fA:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network.go#L157)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

I would legit log these.  glog.Errorf()
</blockquote></details>

Done.

---

*[pkg/kubelet/kubelet_network.go, line 160 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7d-r2-160:-KZ2LOZ2gZLia2iSOCTO:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network.go#L160)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

capture L160-L165 as a function, since it is duplicated above.

```
composedSearch, err := appendOrError(composedSearch, hostDomain)
if err != nil {
  glog...
  return
```

</blockquote></details>

Done.

---

*[pkg/kubelet/kubelet_network_test.go, line 145 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7e-r2-145:-KZ2LPMx_FHPEVzyelJC:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network_test.go#L145)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

All of these would be easier to read it you just used AAA, BBB, CCC for names
</blockquote></details>

Done.

---

*[pkg/kubelet/kubelet_network_test.go, line 148 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7e-r2-148:-KZ2LQ-j1dw1u4wmwUV2:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network_test.go#L148)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

add a simple case with A, B + C, D = ABCD
</blockquote></details>

Done.

---

*[pkg/kubelet/kubelet_network_test.go, line 154 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7e-r2-154:-KZ2LQFuaFv3uvRPBH24:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network_test.go#L154)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

strings.Repeat("a", 100) makes this WAY easier to read
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666)*
<!-- Sent from Reviewable.io -->

vefimova commented:
@thockin PTAL, addressed all comments and implemented requested changes. Please, take a look at PR description (first comment) updated it with the summary list of changes.

thockin commented:
just a couple minor nits and this is golden.  I tagged this as needing a release note, if you don't mind adding one to the first comment (see other recent PRs for examples)

vefimova commented:
@thockin added release note to the first comment

thockin commented:
There are still a handful of very minor comments open

vefimova commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/kubelet/kubelet_network.go, line 125 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666#-KOgmWrHheJx01raCQ7d-r2-125:-KaC3N_gHcYqPS9deqyf:b-7ql1td) ([raw file](https://github.com/kubernetes/kubernetes/blob/eee07ab07fa5f6074075d8b60408e15a4e0dbd6f/pkg/kubelet/kubelet_network.go#L125)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29666)*
<!-- Sent from Reviewable.io -->

vefimova commented:
@thockin 
I suppose you're referring to "code-review/reviewable — 5 files, 8 discussions left (thockin, vishh)"

But all these comment were fixed in recent commit, I added comments like "Done/Fixed" for all of them, but I can't close these discussion thread, afaiu only reviewers involved in discussion can click Acknowledge the discussion thread will be Resolved.

Or am I missing something?

thockin commented:
Sorry, I could see the handful of small comments, but I had not hit "send review", so you didn't see them

vefimova commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +91,115 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+	searchLen := len(strings.Join(composedSearch, " "))
+
+	if searchLen == 0 || (searchLen+len(domain)+1 <= types.MaxSearchLineLen && len(composedSearch) < types.MaxSearchLineDNSDomains) {
+		return true, append(composedSearch, domain)
```
not relevant after re-work

vefimova commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +91,115 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+func wouldFit(kl *Kubelet, pod *v1.Pod, composedSearch []string, domain string) (bool, []string) {
+	searchLen := len(strings.Join(composedSearch, " "))
+
+	if searchLen == 0 || (searchLen+len(domain)+1 <= types.MaxSearchLineLen && len(composedSearch) < types.MaxSearchLineDNSDomains) {
```
not relevant after re-work

vefimova commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +91,115 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+}
+
+func (kl *Kubelet) formDNSSearchForDNSDefault(hostSearch []string, pod *v1.Pod) []string {
+	if len(hostSearch) <= types.MaxSearchLineDNSDomains && len(strings.Join(hostSearch, " ")) <= types.MaxSearchLineLen {
```
re-worked dns search line forming logic, at first just exclude all duplicates using map, after cut it to fit limits if needed.

vefimova commented on pkg/kubelet/types/constants.go:
```diff
@@ -19,4 +19,7 @@ package types
 	// system default DNS resolver configuration
 	ResolvConfDefault = "/etc/resolv.conf"
+	// resolver file Search line current limitations
+	MaxSearchLineDNSDomains = 6
```
there're only two funcs work with resolver limits directly, now these are just vars inside funcs

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1248,6 +1248,11 @@ func (kl *Kubelet) Run(updates <-chan kubetypes.PodUpdate) {
 
+	// Start gorouting responsible for checking limits in resolv.conf
+	if kl.resolverConfig != "" {
+		go wait.Until(func() { kl.checkLimitsForResolvConf() }, 30*time.Second, wait.NeverStop)
```
You say "done" but it is still forking a goroutine?

vefimova replied:
I thought you wanted to trigger node's level event as well on pod's creation, now I see it's not the case, so removed added call from  kl.syncLoopIteration in case of ADD op. Re gorouting forking, see comment below.

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1773,6 +1772,8 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 		switch u.Op {
 		case kubetypes.ADD:
+			// Perform limits check for resolv.conf
+			kl.checkLimitsForResolvConf()
```
What I meant by checking it was that `formDNSSearchFitsLimits` already checks the limits and truncates if needed.  We don't need ANOTHER function that just opens, parses, verifies, logs, and returns, do we?  Or am I missing some reason to do both?

vefimova replied:
My vision of when we should logging and eventing (https://github.com/kubernetes/kubernetes/pull/29666#issuecomment-247620361) was based on one of last comment in discussion https://github.com/kubernetes/kubernetes/pull/29666#issuecomment-240281656:

>A goroutine that monitors the flag-configured resolv.conf for > 3 search
paths and logs an event against the node is uncontroversial.

Monitoring of resolv.conf search line fits limits was an alternative to failure mode (unhealth state for node) which @vishh proposed and he agreed to do in more "safe" way. And kl.checkLimitsForResolvConf() checks only limits, not duplicates. 

So I supposed we'd like to have two type events per node and per pod's, and alarming logs for cluster admins on pod's creation if limits've been exceeded and periodically alarms on node's resolv.conf limits exceeded.

thockin replied:
I see, that's a good distinction that I think I missed.

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +90,123 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+
+	combinedSearch := append(dnsSearch, hostSearch...)
+
+	omitDuplicates(kl, pod, &combinedSearch)
```
rather than pass pointer-to-slice (which is just uncommon in go), can we instead do:

```
combinedSearch = omitDuplicates(kl, pod, combinedSearch)

vefimova replied:
will change it

vefimova replied:
fixed

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +90,120 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+func omitDuplicates(kl *Kubelet, pod *v1.Pod, combinedSearch []string) []string {
+	uniqueDomains := map[string]bool{}
+
+	for _, dnsDomain := range combinedSearch {
```
This is tricky enough that it warrants a unit test.  If this is all I find, I'll approve without it, and you can follow-up

vefimova replied:
I've added unit test for checking return value of formDNSSearch, which consists of omitDuplicates call and formDNSSearchFitsLimits, you think it's not enough?

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +90,120 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+
+	if resolvSearchhLineStrLen := len(strings.Join(composedSearch, " ")); resolvSearchhLineStrLen > resolvSearchLineLenLimit {
+		cutDomainsNum := 0
+		cutDoaminsLen := 0
```
typo: Doamins

vefimova replied:
will fix

thockin commented on pkg/kubelet/kubelet_network.go:
```diff
@@ -89,6 +90,120 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
+	return formDNSSearchFitsLimits(kl, pod, combinedSearch)
+}
+
+func (kl *Kubelet) checkLimitsForResolvConf() {
```
I still feel like this could merge into omitDuplicates() and just send an event to both destinations.  But it's not qorth another back-and-forth

vefimova replied:
I think this resolver.conf check should be done periodically as it more relates to node's status, and this was a compromise of not making node unhealth and exclude it from cluster until the issue exists. Also in case of cluster in more or less stable state, when all pods deployed, such events will be cleared after 1h.

Also I think we can't just add into omitDuplicates() as pods could be added and updated by bunches and it doesn't make sense to check several times at once and  generate bunches of the same events and logs. 

So if you still want to get rid of checking it inside goroutune, then maybe the right place will be to call it inside syncLoopIteration for ADD and Update ops, WDYT?

thockin replied:
OK, I can live with it as a goroutine.  Would be nice to have a function that captured the duplicate logic and logged the appropriate event.  Can you send a followup, since this merged?

Thanks!

thockin reviewed (commented):
LGTM - can you fix up the details in a followup?

vefimova commented:
@k8s-bot verify test this

vefimova commented:
@k8s-bot cvm gke e2e test this
